pallets/gear-builtin-actor/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+                .try_into()
+                .unwrap_or_else(|_| unreachable!("Response message is too large"));
+            let reply_id = MessageId::generate_reply(message_id);
+            let packet = ReplyPacket::new(payload, 0);
+            let dispatch = ReplyMessage::from_packet(reply_id, packet)
+                .into_dispatch(actor_id, origin, message_id);
+
+            journal.push(JournalNote::SendDispatch {
+                message_id,
+                dispatch,
+                delay: 0,
+                reservation: None,
+            });
+
+            let outcome = DispatchOutcome::Success;
+            journal.push(JournalNote::MessageDispatched {
+                message_id,
+                source: origin,
+                outcome,
+            });
+
+            journal.push(JournalNote::MessageConsumed(message_id));
+
+            journal
+        }
+
+        // Error in the actor, generates error reply
+        pub fn process_error(
+            dispatch: &StoredDispatch,
+            gas_spent: u64,
+            err: BuiltinActorError,
+        ) -> Vec<JournalNote> {
+            let message_id = dispatch.id();
+            let origin = dispatch.source();
+            let actor_id = dispatch.destination();
+
+            let mut journal = vec![];
+
+            journal.push(JournalNote::GasBurned {
+                message_id,
+                amount: gas_spent,
+            });
+
+            let err_payload = err
+                .to_string()
+                .into_bytes()
+                .try_into()
+                .unwrap_or_else(|_| unreachable!("Error message is too large"));
+            let err: SimpleExecutionError = err.into();
+
+            let dispatch = ReplyMessage::system(message_id, err_payload, err)
+                .into_dispatch(actor_id, origin, message_id);
+
+            journal.push(JournalNote::SendDispatch {
+                message_id,
+                dispatch,
+                delay: 0,
+                reservation: None,
+            });
+
+            let outcome = DispatchOutcome::MessageTrap {
+                program_id: actor_id,
+                trap: err.to_string(),
+            };
+            journal.push(JournalNote::MessageDispatched {
+                message_id,
+                source: origin,
+                outcome,
+            });
+
+            journal.push(JournalNote::MessageConsumed(message_id));
+
+            journal
+        }
+    }
+}
+
+impl<T: Config> BuiltinRouterProvider<StoredDispatch, JournalNote> for Pallet<T> {
+    type Router = BuiltinRegistry<T>;
+
+    fn provide() -> BuiltinRegistry<T> {
+        BuiltinRegistry::<T>::new()
+    }
+}
+
+pub struct BuiltinRegistry<T: Config> {
+    pub registry: BTreeMap<ProgramId, BuiltinId>,
+    pub _phantom: sp_std::marker::PhantomData<T>,
+}
+impl<T: Config> BuiltinRegistry<T> {
+    fn new() -> Self {
+        let mut registry = BTreeMap::new();
+        let mut builtin_ids = Vec::new();
```

Comment:
It is better to preallocate some space for items. Are we able to use count from an `impl_for_tuples` attribute for that?

Reply:
> It is better to preallocate some space for items. Are we able to use count from an `impl_for_tuples` attribute for that?

We use a hardcoded constant - the max number of elements in a tuple (that is, the max supported number of builtin actors) currently set to 16. Could reuse it here.